File: count_files.py
#!/usr/bin/env python3

import os
import sys
import time
from pathlib import Path
from collections import defaultdict, Counter

class FileCounter:
    def __init__(self):
        self.total_files = 0
        self.total_folders = 0
        self.total_size = 0
        self.extensions = Counter()
        self.errors = []
        
    def count_files(self, folder_path, show_progress=True):
        """Count files recursively using os.walk for memory efficiency"""
        print(f"Scanning folder: {folder_path}")
        print("Press Ctrl+C to stop\n")
        
        start_time = time.time()
        
        try:
            for root, dirs, files in os.walk(folder_path, followlinks=False):
                # Count folders
                self.total_folders += len(dirs)
                
                # Process files
                for file_name in files:
                    self.total_files += 1
                    file_path = os.path.join(root, file_name)
                    
                    try:
                        # Get file size
                        file_size = os.path.getsize(file_path)
                        self.total_size += file_size
                        
                        # Track extensions
                        ext = Path(file_name).suffix.lower() or 'no_extension'
                        self.extensions[ext] += 1
                        
                    except (OSError, PermissionError) as e:
                        self.errors.append(f"Cannot access {file_path}: {str(e)}")
                    
                    # Show progress
                    if show_progress and self.total_files % 1000 == 0:
                        elapsed = time.time() - start_time
                        rate = self.total_files / elapsed if elapsed > 0 else 0
                        print(f"\rScanned {self.total_files:,} files, {self.total_folders:,} folders "
                              f"({rate:.0f} files/sec)", end='', flush=True)
                
        except KeyboardInterrupt:
            print("\n\nScan interrupted by user")
            return None
        except Exception as e:
            print(f"\nError during scan: {e}")
            return None
            
        return self.get_results()
    
    def get_results(self):
        return {
            'total_files': self.total_files,
            'total_folders': self.total_folders,
            'total_size': self.total_size,
            'total_size_gb': self.total_size / (1024**3),
            'extensions': self.extensions,
            'errors': self.errors
        }
    
    def print_results(self, results):
        print('\n' + '='*50)
        print('SCAN RESULTS')
        print('='*50)
        print(f"Total files: {results['total_files']:,}")
        print(f"Total folders: {results['total_folders']:,}")
        print(f"Total size: {results['total_size']/(1024**2):.2f} MB ({results['total_size_gb']:.2f} GB)")
        
        if results['extensions']:
            print('\nTop 10 file extensions:')
            print('-'*30)
            
            for ext, count in results['extensions'].most_common(10):
                percentage = (count / results['total_files']) * 100
                print(f"{ext:<15} {count:>8,} ({percentage:.1f}%)")
        
        if results['errors']:
            print(f"\nErrors encountered: {len(results['errors'])}")
            if len(results['errors']) <= 5:
                for error in results['errors']:
                    print(f"  - {error}")
            else:
                for error in results['errors'][:3]:
                    print(f"  - {error}")
                print(f"  ... and {len(results['errors']) - 3} more errors")

def main():
    if len(sys.argv) < 2:
        print("Usage: python count_files.py <folder-path> [--no-progress]")
        print("Example: python count_files.py ./my-large-folder")
        sys.exit(1)
    
    folder_path = sys.argv[1]
    show_progress = '--no-progress' not in sys.argv
    
    # Check if folder exists
    if not os.path.isdir(folder_path):
        print(f"Error: {folder_path} is not a valid directory")
        sys.exit(1)
    
    counter = FileCounter()
    start_time = time.time()
    
    results = counter.count_files(folder_path, show_progress)
    
    if results:
        end_time = time.time()
        duration = end_time - start_time
        
        counter.print_results(results)
        print(f"\nScan completed in {duration:.2f} seconds")

if __name__ == "__main__":
    main()

File: count-files.js
#!/usr/bin/env node

import { promises as fs } from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

class FileCounter {
  constructor() {
    this.totalFiles = 0;
    this.totalFolders = 0;
    this.totalSize = 0;
    this.extensions = new Map();
    this.errors = [];
    this.progressInterval = 1000;
  }

  async countFiles(folderPath, showProgress = true) {
    console.log(`Scanning folder: ${folderPath}`);
    console.log('Press Ctrl+C to stop\n');

    try {
      await this.scanDirectory(folderPath, showProgress);
    } catch (error) {
      console.error(`Error scanning directory: ${error.message}`);
      return null;
    }

    return this.getResults();
  }

  async scanDirectory(dirPath, showProgress) {
    try {
      const entries = await fs.readdir(dirPath, { withFileTypes: true });
      
      for (const entry of entries) {
        const fullPath = path.join(dirPath, entry.name);
        
        try {
          if (entry.isDirectory()) {
            this.totalFolders++;
            // Recursively scan subdirectories
            await this.scanDirectory(fullPath, showProgress);
          } else if (entry.isFile()) {
            this.totalFiles++;
            
            // Get file stats for size
            try {
              const stats = await fs.stat(fullPath);
              this.totalSize += stats.size;
              
              // Track file extensions
              const ext = path.extname(entry.name).toLowerCase() || 'no extension';
              this.extensions.set(ext, (this.extensions.get(ext) || 0) + 1);
            } catch (statError) {
              // File might have been deleted or permission denied
              this.errors.push(`Cannot stat ${fullPath}: ${statError.message}`);
            }
            
            // Show progress
            if (showProgress && this.totalFiles % this.progressInterval === 0) {
              process.stdout.write(`\rScanned ${this.totalFiles.toLocaleString()} files, ${this.totalFolders.toLocaleString()} folders...`);
            }
          }
        } catch (entryError) {
          this.errors.push(`Cannot access ${fullPath}: ${entryError.message}`);
        }
      }
    } catch (error) {
      this.errors.push(`Cannot read directory ${dirPath}: ${error.message}`);
    }
  }

  getResults() {
    return {
      totalFiles: this.totalFiles,
      totalFolders: this.totalFolders,
      totalSize: this.totalSize,
      totalSizeGB: (this.totalSize / (1024 ** 3)).toFixed(2),
      extensions: this.extensions,
      errors: this.errors
    };
  }

  printResults(results) {
    console.log('\n' + '='.repeat(50));
    console.log('SCAN RESULTS');
    console.log('='.repeat(50));
    console.log(`Total files: ${results.totalFiles.toLocaleString()}`);
    console.log(`Total folders: ${results.totalFolders.toLocaleString()}`);
    console.log(`Total size: ${(results.totalSize / (1024 ** 2)).toFixed(2)} MB (${results.totalSizeGB} GB)`);
    
    if (results.extensions.size > 0) {
      console.log('\nTop 10 file extensions:');
      console.log('-'.repeat(30));
      
      const sortedExtensions = Array.from(results.extensions.entries())
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10);
      
      sortedExtensions.forEach(([ext, count]) => {
        const percentage = ((count / results.totalFiles) * 100).toFixed(1);
        console.log(`${ext.padEnd(15)} ${count.toLocaleString().padStart(8)} (${percentage}%)`);
      });
    }
    
    if (results.errors.length > 0) {
      console.log(`\nErrors encountered: ${results.errors.length}`);
      if (results.errors.length <= 5) {
        results.errors.forEach(error => console.log(`  - ${error}`));
      } else {
        results.errors.slice(0, 3).forEach(error => console.log(`  - ${error}`));
        console.log(`  ... and ${results.errors.length - 3} more errors`);
      }
    }
  }
}

// Command line interface
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.log('Usage: node count-files.js <folder-path> [--no-progress]');
    console.log('Example: node count-files.js ./my-large-folder');
    process.exit(1);
  }
  
  const folderPath = args[0];
  const showProgress = !args.includes('--no-progress');
  
  // Check if folder exists
  try {
    const stats = await fs.stat(folderPath);
    if (!stats.isDirectory()) {
      console.error(`Error: ${folderPath} is not a directory`);
      process.exit(1);
    }
  } catch (error) {
    console.error(`Error: Cannot access ${folderPath} - ${error.message}`);
    process.exit(1);
  }
  
  const counter = new FileCounter();
  const startTime = Date.now();
  
  const results = await counter.countFiles(folderPath, showProgress);
  
  if (results) {
    const endTime = Date.now();
    const duration = ((endTime - startTime) / 1000).toFixed(2);
    
    counter.printResults(results);
    console.log(`\nScan completed in ${duration} seconds`);
  }
}

// Handle Ctrl+C gracefully
process.on('SIGINT', () => {
  console.log('\n\nScan interrupted by user');
  process.exit(0);
});

if (process.argv[1] === __filename) {
  main().catch(console.error);
}

export default FileCounter;